pkg/codegen/internal/test/testdata/simple-enum-schema/types/enums.ts

```diff
@@ -0,0 +1,25 @@
+// *** WARNING: this file was generated by test. ***
+// *** Do not edit by hand unless you're certain you know what you are doing! ***
+
+import * as pulumi from "@pulumi/pulumi";
+
+export const redContainerColor: ContainerColor = "red";
+export const blueContainerColor: ContainerColor = "blue";
+export const yellowContainerColor: ContainerColor = "yellow";
+
+export type ContainerColor = "red" | "blue" | "yellow";
+
+export const FourInchContainerSize: ContainerSize = 4;
+export const SixInchContainerSize: ContainerSize = 6;
+export const EightInchContainerSize: ContainerSize = 8;
+
+export type ContainerSize = 4 | 6 | 8;
+
+/** A burgundy rubber tree. */
+export const BurgundyRubberTreeVariety: RubberTreeVariety = "Burgundy";
+/** A ruby rubber tree. */
+export const RubyRubberTreeVariety: RubberTreeVariety = "Ruby";
+/** A tineke rubber tree. */
+export const TinekeRubberTreeVariety: RubberTreeVariety = "Tineke";
+
+export type RubberTreeVariety = "Burgundy" | "Ruby" | "Tineke";
```

Comment:
Sorry for not commenting on this in the design doc earlier, but it only occurred to me just now...

I know we've typically emitted these in the past like what's currently being done here, but I was wondering if we should use [`const enums`](https://www.typescriptlang.org/docs/handbook/enums.html#const-enums) instead:

```typescript
export const enum ContainerColor {
    Red = "red",
    Blue = "blue",
    Yellow = "yellow",
}

export const enum ContainerSize {
    FourInch = 4,
    SixInch = 6,
    EightInch = 8,
}

export const enum RubberTreeVariety {
    /** A burgundy rubber tree. */
    Burgundy = "Burgundy",
    /** A ruby rubber tree. */
    Ruby = "Ruby",
    /** A tineke rubber tree. */
    Tineke = "Tineke",
}
```

Unfortunately, there appears to be a downside: the value would have to be specified as `RubberTreeVariety.Burgundy` rather than `"Burgundy"`.

```typescript
export interface RubberTreeArgs {
    readonly type?: RubberTreeVariety;
}

const a: RubberTreeArgs = {
    type: RubberTreeVariety.Burgundy, // Works
};

const b: RubberTreeArgs = {
    type: "Burgundy", // Doesn't work
};
```

To solve this, I was wondering if there was a built-in way in TypeScript to lookup the type of the enum property values, but there doesn't look like there's a way to do that, so we'd have to do something like:

```typescript
export const enum RubberTreeVariety {
    /** A burgundy rubber tree. */
    Burgundy = "Burgundy",
    /** A ruby rubber tree. */
    Ruby = "Ruby",
    /** A tineke rubber tree. */
    Tineke = "Tineke",
}

export type RubberTreeVarietyValues = "Burgundy" | "Ruby" | "Tineke";

export interface RubberTreeArgs {
    readonly type?: RubberTreeVariety | RubberTreeVarietyValues;
}
```

But having to have different named types (`RubberTreeVariety` and `RubberTreeVarietyValues`) is not ideal. So not really loving that.

However, this exploration led me to the following blog post: https://www.designcise.com/web/tutorial/is-it-possible-to-allow-literal-string-values-with-typeScripts-enum-type

I think we should consider doing something along the lines of that post:

```typescript
export const ContainerColor = {
    Red: "red",
    Blue: "blue",
    Yellow: "yellow",
} as const;

export type ContainerColor = (typeof ContainerColor)[keyof typeof ContainerColor]; // "red" | "blue" | "yellow"

export const ContainerSize = {
    FourInch: 4,
    SixInch: 6,
    EightInch: 8,
} as const;

export type ContainerSize = (typeof ContainerSize)[keyof typeof ContainerSize]; // 4 | 6 | 8

export const RubberTreeVariety = {
    /** A burgundy rubber tree. */
    Burgundy: "Burgundy",
    /** A ruby rubber tree. */
    Ruby: "Ruby",
    /** A tineke rubber tree. */
    Tineke: "Tineke",
} as const;

export type RubberTreeVariety = (typeof RubberTreeVariety)[keyof typeof RubberTreeVariety]; // "Burgundy" | "Ruby" | "Tineke"
```

It allows both of these uses:

```typescript
// Both work.
const a: RubberTreeArgs = {
    type: RubberTreeVariety.Burgundy, // Works
};
const b: RubberTreeArgs = {
    type: "Burgundy", // Works
};
```

And doesn't require two different type names.

And provides a cleaner way of grouping the constants together, improving discoverability, etc.

Reply:
Interesting find! I agree that it's cleaner / more discoverable. It's not immediately clear to me if this would play nice with the existing enums in pulumi-aws where the constants follow a `BurgundyRubberTreeVariety` pattern and are already in use, so this would be a breaking change. I'll spend some time looking into whether/how we could support both.

Reply:
Okay, I've created an issue for this https://github.com/pulumi/pulumi/issues/5494 and definitely would like to see us make this improvement in the future. For now, I'm going to continue with the existing implementation as it matches the enums we already have in the wild (pulumi-aws), and would like to focus on schematizing what we have and moving on to the other languages.